core/controllers/load_functions.py

```diff
@@ -0,0 +1,257 @@
+        matching_project = None
+        for project in programme_exists_same_round.projects:
+            if project.submission.reporting_round == reporting_round:
+                matching_project = project
+                break
+        if matching_project:
+            # get id of submission to replace and re-use it
+            submission_id = matching_project.submission.submission_id
+            # submission instance to remove
+            submission_to_del = matching_project.submission
+            # drop submission, all children should be dropped via cascade
+            Submission.query.filter_by(id=submission_to_del.id).delete()
+            db.session.flush()
+
+    else:
+        submission_id = next_submission_id(reporting_round)
+
+    return submission_id
+
+
+def next_submission_id(reporting_round: int) -> str:
+    """Get the next submission ID by incrementing the last in the DB.
+
+    :return: The next submission ID.
+    """
+    # Conversion to cast numpy int types from pandas data extract
+    reporting_round = int(reporting_round)
+    latest_submission = (
+        Submission.query.filter_by(reporting_round=reporting_round)
+        # substring submission number digits, cast to int and order to get the latest submission
+        .order_by(desc(func.cast(func.substr(Submission.submission_id, 7), db.Integer))).first()
+    )
+    if not latest_submission:
+        return SUBMISSION_ID_FORMAT.format(reporting_round, 1)  # the first submission
+
+    incremented_submission_num = latest_submission.submission_number + 1
+    return SUBMISSION_ID_FORMAT.format(reporting_round, incremented_submission_num)
+
+
+def get_outcomes_outputs_to_insert(mapping: DataMapping, models: list) -> list:
+    """Returns outcomes or outputs not present in the database.
+
+    :param mapping: mapping of ingest to db
+    :param models: list of incoming outcomes or outputs being ingested
+    :return: list of outcomes or outcomes to be inserted
+    """
+    db_model_field = {"Outputs_Ref": "output_name", "Outcome_Ref": "outcome_name"}[mapping.table]
+
+    query_results = db.session.query(getattr(mapping.model, db_model_field)).all()
+    existing_names = [str(row[0]) for row in query_results]
+    models = [model for model in models if getattr(model, db_model_field) not in existing_names]
+
+    return models
+
+
+def remove_unreferenced_organisations():
+    """Removes organisations no longer referenced by a Programme.
+
+    Some organisation names have been wrong in previous ingest.
+    This will clean them up everytime a new ingest is run updating those names
+    """
+
+    organisations_without_child = Organisation.query.filter(~Organisation.programmes.any()).all()
+    if organisations_without_child:
+        for organisation in organisations_without_child:
+            db.session.delete(organisation)
+
+    db.session.commit()
+
+
+def get_programmes_same_round_or_older(reporting_round: int, programme_ids: list[str]) -> dict[str, str]:
```

Comment:
I wonder, should we split these historical ingest methods into a different module, something like ingest_historical or load_functions_historical? Just to keep them separate from the core ingest code? 

Might make it easier for people to mentally de-couple the different ingest methods going forwards

Reply:
Yes, probably a good idea